feat(notifications): Build robust notification center

sangeethsuresh123 opened the pull request:
### Purpose

This PR introduces the new **Notification Center**, a feature-rich module for displaying user notifications. The implementation serves as a blueprint for building scalable, type-safe, and robust features using the project's modern tech stack.

---
### Key Changes & Refinements

* **State Management**: Fully implemented using **TanStack Query** for all server state, caching, and mutations, providing a resilient and efficient data-fetching layer.

* **Service-Layer Architecture**: All API logic is encapsulated in a dedicated `notificationService`, cleanly separating data-fetching concerns from the UI.

* **Runtime Type Safety**: Integrated **Zod** to define strict data schemas. This guarantees that all data from the API (or mock service) is validated at runtime, preventing bugs caused by unexpected data shapes.

* **Robust Error Handling**: The `useNotifications` hook includes comprehensive error handling. It safely catches and reports both API errors and Zod validation failures, preventing crashes and providing clear feedback in the UI. All ESLint type-safety rules have been resolved.

* **Code Conventions**: The feature follows project standards with a clean separation of UI components, hooks, services, and types. An isolated `NotificationsProvider` is used to scope TanStack Query to this feature.

---
### How to Test

1.  Navigate to the `/notifications` route.
2.  Verify the loading state appears, followed by a grid of notification cards.
3.  Click "Mark as Read" on a card. Confirm the button enters a pending state and the card is removed on success.

### Screenshots
<img width="3192" height="1668" alt="pic1" src="https://github.com/user-attachments/assets/be4af55d-0c10-4ed7-a1c8-02f9b0558213" />
<img width="3196" height="1660" alt="image" src="https://github.com/user-attachments/assets/bac5c52e-0b36-4824-a16a-4a0b86777ba3" />
<img width="3200" height="1662" alt="pic3" src="https://github.com/user-attachments/assets/f03f666d-2c74-4611-852a-59fe17c366df" />

sangeethsuresh123 committed:
chore(deps): update dependencies

sangeethsuresh123 committed:
feat(ui): add avatar component for user display

sangeethsuresh123 committed:
fix(sidebar): correct vertical alignment of back button

Refactored the component to use a flexbox layout (justify-between) to prevent the back button from overlapping navigation items.

sangeethsuresh123 committed:
feat(notifications): implement notification center page

Adds the main page component, Zustand store, and UI for the new Notification Center. This page displays system messages and alerts to the user.

alpha-og reviewed (changes requested):
Provide screenshots of the introduced feature including different component states and/ or a demo video showcasing the changes made.

sangeethsuresh123 committed:
refactor(notifications): separate feature logic into services and hooks

Refactored the notification feature to follow the service-layer architecture. This addresses PR feedback by separating the state, logic, and data-fetching into their respective files for better maintainability and testability.

sangeethsuresh123 committed:
refactor(notifications): separate feature logic into services and hooks

Refactored the notification feature to follow the service-layer architecture. This addresses PR feedback by separating the state, logic, and data-fetching into their respective files for better maintainability and testability.

sangeethsuresh123 committed:
refactor(notifications): separate feature logic into services and hooks

Refactored the notification feature to follow the service-layer architecture. This addresses PR feedback by separating the state, logic, and data-fetching into their respective files for better maintainability and testability.

sangeethsuresh123 committed:
feat(query): set up TanStack Query provider

Adds and configures the QueryClientProvider in a central providers component. This is required for TanStack Query to function across the application.

sangeethsuresh123 committed:
refactor(notifications): migrate state management to TanStack Query

Replaces the Zustand-based state management with TanStack Query for server state. This refactor includes:
A Zod schema for type validation.

A service layer using Axios.

A custom hook with useQuery and useMutation.

sangeethsuresh123 committed:
refactor(notifications): update page component to use TanStack Query hook

Updates the NotificationsPage to consume the new useNotifications hook. The component now gets its state (data, loading, error) directly from TanStack Query and handles mutation pending states for a better user experience.

sangeethsuresh123 committed:
refactor(notifications): migrate state management to TanStack Query

Replaces the Zustand-based state management with TanStack Query. This includes a Zod schema, a service layer, and a hook using useQuery/useMutation.

RoshinR2005 commented on apps/client-web/src/features/notifications/services/notificationService.ts:
Seperate the mock data from the services and comment the real api request

sangeethsuresh123 committed:
feat(notifications): create isolated provider for TanStack Query

Adds a feature-specific provider to manage the QueryClient, isolating it from the global scope as required.

sangeethsuresh123 committed:
refactor(notifications): introduce service layer with mock data

Separates mock data from the service logic and deletes the old Zustand store. The new service file includes commented-out real API calls for future integration.

sangeethsuresh123 committed:
refactor(notifications): connect page to hook and provider

Updates the NotificationsPage to use the new isolated provider and consume the useNotifications hook. The UI now handles loading, error, and pending mutation states correctly.

sangeethsuresh123 committed:
revert(providers): restore original providers file

Reverts changes to the global providers.tsx file to avoid side-effects. The feature will use an isolated provider instead.

JVP555 reviewed (changes requested):
1. Move the current component from notifications/page.tsx to a new folder: features/notifications/components, dedicated for notification-related components.

2. Update page.tsx to properly import and export the relocated component.

3. Refactor the component to use type-based structures instead of interfaces, following our current type conventions.

4. Update the folder structure and service imports according to the paths and conventions outlined in the documentation shared on Discord.

sangeethsuresh123 committed:
feat(notifications): define final notification schema

Establishes the Zod schema for the Notification and NotificationPriority types. Uses the modern z.guid() and z.iso.datetime() for strict, safe validation.

sangeethsuresh123 committed:
feat(notifications): implement mock data and service layer

Adds a mock database with correctly formatted data that passes the Zod schema validation. The service file provides functions to get and modify this mock data.

sangeethsuresh123 committed:
feat(notifications): create useNotifications hook

Implements the core logic hook for the feature. It uses TanStack Query to fetch data, safely parses it with the Zod schema, and provides state and mutation functions to the UI.

sangeethsuresh123 committed:
feat(notifications): build notification center UI

Creates the NotificationView component to display the list of notifications. The page file and isolated provider are set up to render the component correctly.

JVP555 committed:
Merge branch 'frontend' into feature/notification-center

KK-8848 committed:
update:updated the services and switched form mock to actual data implementation

KK-8848 committed:
Merge branch 'frontend' into feature/notification-center
